### PR TITLE
Selectively replace images/containers in a Data Gate pod

### PR DESCRIPTION
**Pull request by johker01**
This pull request allows replacing specific images in the Data Gate pod for testing purposes, e.g. when wanting to run Data Gate test cases in a CI/CD pipeline. It is possible to either supply a deployment name or if not supplied, the CLI command will try to automatically derive the deployment name. This will not work, if more than one deployment is running on the same cluster.
This is a synchronous call, meaning that the command will only return when the image replacement is done and all containers in the pod are running. The mechanism how to check whether the pod has been restarted can and will be improved in a future pull request, as it is done [here](https://github.com/IBM/cloud-pak-operations-cli/blob/master/cpo/lib/cloud_pak_for_data/cpd_4_0_0/cpd_manager.py#L1366).

### CODE REVIEW

File: cpo/lib/openshift/oc.py [hunk 1]
```diff
@@ -226,3 +228,74 @@ def log_in_to_openshift_cluster_with_token(server: str, token: str):
...
+
+    if not result:
+        raise DataGateCLIException(f"Pod(s) containing the string '{search_string}' could not be found")
+
+    return result
+
+
+def get_pod_status(pod_name: str) -> GetPodEntry:
+    """Returns the current pod status for a given pod name"""
+    oc_get_pods_args = ["get", "pods"]
```
Comment (jensmueller-com) [inline cpo/lib/openshift/oc.py:271]: Either return and parse JSON (`oc get pods --output json`) or use JSONPath (`oc get pods {pod_name} --output jsonpath="{.status.phase}"`)

File: cpo/lib/openshift/oc.py [hunk 2]
```diff
@@ -226,3 +228,74 @@ def log_in_to_openshift_cluster_with_token(server: str, token: str):
...
+        if pod_name in line:
+            break
+
+    return GetPodEntry.parse(line)
+
+
+def get_custom_resource(custom_resource_kind: str, custom_resource_id: str) -> dict:
+    """Get the custom resource for a given kind and id"""
+
+    oc_get_custom_resource_args = ["get", custom_resource_kind, custom_resource_id, "-o", "json"]
```
Comment (jensmueller-com) [inline cpo/lib/openshift/oc.py:286]: Use long command line options (see [coding guidelines](https://github.com/IBM/cloud-pak-operations-cli/blob/master/docs/coding_guidelines.md#command-line-options))

File: cpo/lib/openshift/oc.py [hunk 3]
```diff
@@ -226,3 +228,74 @@ def log_in_to_openshift_cluster_with_token(server: str, token: str):
...
+    line = ""
+    for line in oc_get_pods_result.splitlines():
+        if pod_name in line:
+            break
+
+    return GetPodEntry.parse(line)
+
+
+def get_custom_resource(custom_resource_kind: str, custom_resource_id: str) -> dict:
+    """Get the custom resource for a given kind and id"""
```
Comment (jensmueller-com) [inline cpo/lib/openshift/oc.py:284]: "id" → "ID"

File: cpo/lib/openshift/types/get_pod_entry.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+#  Copyright 2022 IBM Corporation
```
Comment (jensmueller-com) [inline cpo/lib/openshift/types/get_pod_entry.py:1]: Remove and either return and parse JSON (`oc get pods --output json`) or use JSONPath (`oc get pods {pod_name} --output jsonpath="{.status.phase}"`) [see `cpo/lib/openshift/oc.py`]

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+        raise DataGateCLIException(f"Unable to retrieve pod name for Data Gate deployment '{deployment_name}'")
+
+    return result
+
+
+def _is_pod_ready(pod_name: str):
+    return oc.get_pod_status(pod_name).is_ready()
+
+
+def _is_pod_running_from_deployment(deployment_name: str):
```
[the remark this answers is not in the document]
Comment (jensmueller-com) [reply cpo/commands/cluster/replace_data_gate_images.py:146]: Add return type

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+    try:
+        pod_name = _get_data_gate_pod_name_from_deployment_name(deployment_name)
+        result = oc.get_pod_status(pod_name).is_running()
+    except Exception:
+        pass
+
+    return result
+
+
+def _is_pod_not_running_from_deployment(deployment_name: str):
```
[the remark this answers is not in the document]
Comment (jensmueller-com) [reply cpo/commands/cluster/replace_data_gate_images.py:158]: Add return type

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+""",
+)
+@click.option("--container-stunnel", help="container-stunnel image tag")
+@click.option("--container-stunnel-init", help="container-stunnel-init image tag")
+@click.option("--data-gate-apply", help="data-gate-apply image tag")
+@click.option("--data-gate-api", help="data-gate-api image tag")
+@click.option("--data-gate-init", help="data-gate-init image tag")
+@click.option("--data-gate-server", help="data-gate-server image tag")
+@click.option("--data-gate-ui", help="data-gate-ui image tag")
+def replace_data_gate_images(
```
Comment (jensmueller-com) [inline cpo/commands/cluster/replace_data_gate_images.py:45]: "data_gate_images" → "db2_data_gate_images"

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    if m and m.group(1):
+        result = m.group(1)
+        logging.debug(f"App name: {result}")
+    else:
+        raise DataGateCLIException(f"Unable to retrieve app name for Data Gate deployment name '{deployment_name}'")
+
+    return result
+
+
+def _get_data_gate_pod_name_from_deployment_name(deployment_name: str) -> str:
```
Comment (jensmueller-com) [inline cpo/commands/cluster/replace_data_gate_images.py:128]: "_get_data_gate_pod_name_from_deployment_name" → "_get_db2_data_gate_pod_name_from_deployment_name"

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        result = app_name.replace("-", "")
+    except DataGateCLIException:
+        raise DataGateCLIException(
+            f"Unable to retrieve custom resource id for Data Gate deployment name '{deployment_name}'"
+        )
+
+    return result
+
+
+def _replace_image_in_data_gate_instance_service_custom_resource(
```
Comment (jensmueller-com) [inline cpo/commands/cluster/replace_data_gate_images.py:183]: "_replace_image_in_data_gate_instance_service_custom_resource" → "_replace_image_in_db2_data_gate_instance_service_custom_resource"

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    timeout: str,
+    container_stunnel: str,
+    container_stunnel_init: str,
+    data_gate_apply: str,
+    data_gate_api: str,
+    data_gate_init: str,
+    data_gate_server: str,
+    data_gate_ui: str,
+):
+    """Replace a single or multiple specific component(s) of a given deployment.
```
[the remark this answers is not in the document]
Comment (jensmueller-com) [reply cpo/commands/cluster/replace_data_gate_images.py:57]: "given deployment" → "given Db2 Data Gate deployment"

**Review by jensmueller-com — CHANGES_REQUESTED** [review]

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 11]
```diff
@@ -0,0 +1,197 @@
+#  Copyright 2021 IBM Corporation
```
[the remark this answers is not in the document]
Comment (johker01) [reply cpo/commands/cluster/replace_data_gate_images.py:1]: done

File: cpo/commands/cluster/replace_data_gate_images.py [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        components["datagate_api"] = data_gate_api
+    if data_gate_init:
+        components["datagate_init"] = data_gate_init
+    if data_gate_server:
+        components["datagate_server"] = data_gate_server
+    if data_gate_ui:
+        components["datagate_ui"] = data_gate_ui
+
+    dg_id = _infer_custom_resource_id_from_deployment_name(deployment_name)
+    custom_resource_json = oc.get_custom_resource("datagateinstanceservice", dg_id)
```
[the remark this answers is not in the document]
Comment (johker01) [reply cpo/commands/cluster/replace_data_gate_images.py:81]: understood

**Review by jensmueller-com — APPROVED** [review]

**Pull request merged**